File: src/main/java/eu/codetopic/utils/ui/activity/modular/ActivityCallBackModule.java
/*
 * Copyright 2017 Jiří Kuchyňka (Anty)
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package eu.codetopic.utils.ui.activity.modular;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.Toolbar;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

public abstract class ActivityCallBackModule {

    private static final String LOG_TAG = "ActivityCallBackModule";

    private ModularActivity mActivity = null;

    final void init(@NonNull ModularActivity activity) {
        if (mActivity != null) throw new IllegalStateException(LOG_TAG + " is still initialized");
        mActivity = activity;
    }

    public ModularActivity getActivity() {
        return mActivity;
    }

    protected abstract void onCreate(@Nullable Bundle savedInstanceState);

    protected abstract void onNewIntent(Intent intent);

    protected abstract void onSetContentView(@LayoutRes int layoutResID,
                                             SetContentViewCallBack callBack);

    protected abstract void onSetContentView(View view, SetContentViewCallBack callBack);

    protected abstract void onSetContentView(View view, ViewGroup.LayoutParams params,
                                             SetContentViewCallBack callBack);

    @Nullable
    protected abstract Toolbar onSetSupportActionBar(@Nullable Toolbar toolbar);

    protected abstract void onPostSetSupportActionBar(@Nullable Toolbar toolbar);

    protected abstract void onRestoreInstanceState(Bundle savedInstanceState);

    protected abstract void onStart();

    protected abstract void onResume();

    protected abstract boolean onCreateOptionsMenu(Menu menu);

    protected abstract boolean onOptionsItemSelected(MenuItem item);

    protected abstract void onTitleChanged(CharSequence title, int color);

    protected abstract void onConfigurationChanged(Configuration newConfig);

    protected abstract void onBackPressed();

    protected abstract boolean onKeyDown(int keyCode, KeyEvent event);

    protected abstract void onAttachFragment(Fragment fragment);

    protected abstract void onSaveInstanceState(Bundle outState);

    protected abstract void onPause();

    protected abstract void onStop();

    protected abstract void onDestroy();

    protected abstract void onRestart();

    public interface SetContentViewCallBack {

        void set(@LayoutRes int layoutResID);

        void set(View view);

        void set(View view, ViewGroup.LayoutParams params);

        void pass();

        void addViewAttachedCallBack(Runnable callBack);

    }
}

File: src/main/java/eu/codetopic/utils/ui/activity/modular/ActivityCallBackModule.kt
/*
 * utils
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package eu.codetopic.utils.ui.activity.modular

import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.v4.app.Fragment
import android.support.v7.widget.Toolbar
import android.view.KeyEvent
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup

abstract class ActivityCallBackModule {

    companion object {

        private const val LOG_TAG = "ActivityCallBackModule"
    }

    lateinit var activity: ModularActivity
        private set

    internal fun init(activity: ModularActivity) {
        if (this::activity.isInitialized) throw IllegalStateException("$LOG_TAG is still initialized")
        this.activity = activity
    }

    abstract fun onCreate(savedInstanceState: Bundle?)

    abstract fun onNewIntent(intent: Intent)

    abstract fun onSetContentView(@LayoutRes layoutResID: Int,
                                  callBack: SetContentViewCallBack)

    abstract fun onSetContentView(view: View, callBack: SetContentViewCallBack)

    abstract fun onSetContentView(view: View, params: ViewGroup.LayoutParams,
                                  callBack: SetContentViewCallBack)

    abstract fun onSetSupportActionBar(toolbar: Toolbar?): Toolbar?

    abstract fun onPostSetSupportActionBar(toolbar: Toolbar?)

    abstract fun onRestoreInstanceState(savedInstanceState: Bundle)

    abstract fun onStart()

    abstract fun onResume()

    abstract fun onCreateOptionsMenu(menu: Menu): Boolean

    abstract fun onOptionsItemSelected(item: MenuItem): Boolean

    abstract fun onTitleChanged(title: CharSequence, color: Int)

    abstract fun onConfigurationChanged(newConfig: Configuration)

    abstract fun onBackPressed()

    abstract fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean

    abstract fun onAttachFragment(fragment: Fragment)

    abstract fun onSaveInstanceState(outState: Bundle)

    abstract fun onPause()

    abstract fun onStop()

    abstract fun onDestroy()

    abstract fun onRestart()

    interface SetContentViewCallBack {

        fun set(@LayoutRes layoutResID: Int)

        fun set(view: View)

        fun set(view: View, params: ViewGroup.LayoutParams)

        fun pass()

        fun addViewAttachedCallBack(callBack: () -> Unit)
    }
}
